app/uk/gov/hmrc/constructionindustryscheme/services/clientlist/ClientListService.scala

```diff
@@ -0,0 +1,207 @@
+                  )
+                  clearWaitTime(credentialId)
+                  Future.failed(SystemException("Initiate on final business interval"))
+          yield result
+
+      for
+        businessSucceeded <- loop(0)
+        outcome <-
+          // BROWSER PHASE: only when we exhaust business intervals with last status = InProgress
+          if businessSucceeded then
+            Future.unit
+          else
+            for
+              _      <- sleep(waitPlan.browserIntervalMs)
+              status <- getStatus(credentialId)
+              _       = logStatus("browser", None, status)
+              finalResult <- status match
+                case Succeeded =>
+                  clearWaitTime(credentialId)
+                  Future.unit
+
+                case Failed =>
+                  audit.clientListRetrievalFailed(credentialId, phase = "browser")
+                  clearWaitTime(credentialId)
+                  Future.failed(ClientListDownloadFailedException("Failed after browser interval"))
+
+                case InitiateDownload =>
+                  audit.clientListRetrievalFailed(credentialId, phase = "browser", reason = Some("initiate-after-browser"))
+                  clearWaitTime(credentialId)
+                  Future.failed(SystemException("Initiate after browser interval"))
+
+                case InProgress =>
+                  audit.clientListRetrievalInProgress(credentialId, phase = "browser")
+                  clearWaitTime(credentialId)
+                  Future.failed(ClientListDownloadInProgressException("Still in progress"))
+            yield finalResult
+      yield outcome
+
+  // ------------------------------------------------------------
+  // Entry point
+  // ------------------------------------------------------------
+
+  def process(credentialId: String)(implicit hc: HeaderCarrier): Future[Unit] =
+    getStatus(credentialId).flatMap {
+      case Succeeded =>
+        logStatus("initial", None, Succeeded)
+        clearWaitTime(credentialId)
+        Future.unit
+
+      case Failed =>
+        logStatus("initial", None, Failed)
+        audit.clientListRetrievalFailed(credentialId, phase = "initial")
+        clearWaitTime(credentialId)
+        Future.failed(ClientListDownloadFailedException("Failed"))
+
+      case InProgress =>
+        logStatus("initial", None, InProgress)
+        audit.clientListRetrievalInProgress(credentialId, phase = "initial")
+        val waitPlan = getCachedWaitTime(credentialId).getOrElse(defaultWaitPlan)
+        processWithWaitPlan(credentialId, waitPlan)
+
+      case InitiateDownload =>
+        logStatus("initial", None, InitiateDownload)
+
+        for
+          waitPlan <- clientExchangeProxyConnector.initiate(serviceName, credentialId)
+          _         = cacheWaitTime(credentialId, waitPlan)
+          outcome  <- processWithWaitPlan(credentialId, waitPlan)
+        yield outcome
+    }
```

Comment:
Interesting implementation. I will spend some more time understanding this, but I have two questions right now.
* Is this relying on some kind of in-memory cache? what happens if we have multiple instances of the service? could we end up making more calls to client list proxy than the legacy app does? If so please check with Vamsi if this is okay...
* Wouldn't it be better to return the status in the service rather than using unit/failed states on the future?

Reply:
In the stub, the service immediately changes from InitiateDownload to InProgress, but according to confluence this state change can take some time. which is why we have the service intervals...

So if it takes say 4 seconds to go from initiate download, and the frontend makes more than one request to get the status while it is InitiateDownload, we could end up making multiple requests to clientlistproxy to initiate the download (if each request is routed to a different instance.

Regardless of wether this is an issue or not (it's probably not, hopefully) I think we should go with this for now, and refactor later if needed.

@jassalrichy are you happy with this approach?